goth/runner/__init__.py

```diff
@@ -324,6 +266,73 @@ async def _exit(self):
         payment.clean_up()
 
 
+@asynccontextmanager
+async def _run_compose_network(
+    compose_manager: ComposeNetworkManager, log_dir: Path, force_build: bool = False
+) -> None:
+    """Implement AsyncContextManager for starting/stopping compose network manager."""
+
+    try:
+        await compose_manager.start_network(log_dir, force_build)
+        yield
+    finally:
+        await compose_manager.stop_network()
+
+
+@contextmanager
+def _create_probe(
+    runner: "Runner",
+    docker_client: docker.DockerClient,
+    config: YagnaContainerConfig,
+    log_config: LogConfig,
+) -> Probe:
+    """Implement a ContextManager protocol for creating and removing probes."""
+
+    probe: Optional[Probe] = None
+    try:
+        probe = config.probe_type(runner, docker_client, config, log_config)
+        for name, value in config.probe_properties.items():
+            probe.__setattr__(name, value)
+        yield probe
+    finally:
+        if probe:
+            probe.remove()
+
+
+@asynccontextmanager
+async def _run_probe(probe: Probe) -> str:
+    """Implement AsyncContextManager for starting and stopping a probe."""
+
+    try:
+        await probe.start()
+        assert probe.ip_address
+        yield probe.ip_address
+    finally:
+        await probe.stop()
+
+
+@contextmanager
+def _run_proxy(proxy: Proxy) -> Proxy:
+    """Implement ContextManager protocol for running a Proxy."""
+
+    try:
+        proxy.start()
+        yield
+    finally:
+        proxy.stop()
+
+
+@asynccontextmanager
+async def _run_web_server(server: WebServer, server_address: Optional[str]) -> None:
+    """Implement AsyncContextManager protocol for starting/stopping a web server."""
+
+    try:
+        await server.start(server_address)
+        yield
+    finally:
+        await server.stop()
+
+
```

Comment:
Is it be better to have all the (Async)ContextManagers here, or in the respective modules (`compose_manager`, `probe` etc.)? I think the context managers should be standalone functions, not methods of respective classes ('ComposeNetworkManager`, `Probe`), to keep class interfaces clean (for example, `Proxy` should have just `start()` and `stop()`, not `start()`, `stop()` and `run()`) but perhaps they belong to the modules where the classes are defined?